polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp

```diff
@@ -478,33 +478,40 @@ static bool canBeHoisted(LICMCandidate &candidate, LoopLikeOpInterface loop,
   if (isMemoryEffectFree(&op)) {
     LLVM_DEBUG({
       llvm::dbgs() << "Operation: " << op << "\n";
-      llvm::dbgs().indent(2) << "**** can be hoisted: has no side effects\n\n";
+      llvm::dbgs().indent(2) << "**** has no side effects\n\n";
     });
-    return true;
+  } else {
```

Comment:
The block at lines 479-482 doesn't do anything anymore (other than printing a trace). We could remove the check `isMemoryEffectFree` (so the if ... then ...else) and hoist the code currently in the else block.





if (!isMemoryEffectFree(&op)) {
    // Do not hoist operations that allocate a resource
    ...
}

Reply:
IIUC, `IsIsolatedFromAbove` doesn't ensure there is no memory operations reference memories from above. 
So, the early exit check at line 478 should be `if (isMemoryEffectFree(&op) && (op.getRegions().empty() || op.hasTrait<OpTrait::IsIsolatedFromAbove>()))`.

Reply:
but the operations going through the `else` branch should not early exit, as  `IsIsolatedFromAbove` doesn't ensure there is no memory operations in the region reference memories from above the region.

Reply:
True, reintroduced. Should be fine now.